Packs/PingCastle/Integrations/PingCastle/PingCastle.py

```diff
@@ -0,0 +1,178 @@
+        self.send_response(code)
+        self.send_header('Content-type', content_type)
+        self.end_headers()
+
+    def do_GET(self):
+        """
+        Handle GET requests. This server does not accept GET requests.
+        """
+        self._set_response(400, 'text/html')
+        self.wfile.write(b"no GET requests")
+
+    def do_POST(self):
+        """
+        Handle POST requests - Login and SendReport
+        """
+        try:
+            content_length = int(self.headers.get('Content-Length'))
+            post_data = self.rfile.read(content_length)
+            try:
+                request_json = json.loads(post_data)
+                if self.path == '/api/Agent/Login':
+                    self._handle_login(request_json)
+
+                elif self.path == '/api/Agent/SendReport':
+                    self._handle_send_report(request_json)
+
+                else:
+                    self._set_response(404, 'text/html')
+                    self.wfile.write(b'no such url')
+
+            except json.JSONDecodeError:
+                self._set_response(400, 'text/html')
+                self.wfile.write(b'invalid format')
+        except Exception:
+            demisto.error(traceback.format_exc())
+
+    def _handle_send_report(self, request_json: dict):
+        """
+        Handle receiving a report from PingCastle making sure the token used is the token issued by this server.
+        Args:
+            request_json (dict): The JSON sent in the request.
+        """
+        if self.headers.get('Authorization') is None:
+            self._set_response(400, 'text/html')
+            self.wfile.write(b'missing token')
+            return
+
+        if self.headers.get('Authorization') != self.token:
+            self._set_response(401, 'text/html')
+            self.wfile.write(b'Invalid Token')
+            return
+
+        if request_json.get('xmlReport') is None:
+            self._set_response(400, 'text/html')
+            self.wfile.write(b'missing xmlReport')
+            return
+
+        self._set_report_context(request_json)
+        self._create_incident(request_json)
+        self._set_response(200, 'text/html')
+
+    def _create_incident(self, request_json: dict):
+        """
+        Create an Incident from a report sent by PingCastle.
+        Args:
+            request_json (dict): The JSON sent in the request.
+        """
+        incident = {
+            'name': 'PingCastle-report',
+            'occurred': time.strftime('%Y-%m-%dT%H:%M:%SZ'),
+            'rawJSON': json.dumps(request_json)
+        }
+        demisto.createIncidents([incident])
+
+    def _set_report_context(self, request_json: dict):
+        """
+        Store the report in the integration context so that it may be retrieved by the get-report command.
+        Args:
+            request_json (dict): The JSON sent in the request.
+        """
+        context: dict = demisto.getIntegrationContext()
```

Comment:
We get the context, set a value inside it then set the result as the new context because if there is anything other than the report in the context we would overwrite it if we didn't do it this way.